File: rust/src/valid_palindrome_125.rs
use crate::Solution;
//NOTE: Double ended iterators are pretty awesome
impl Solution {
    pub fn is_palindrome_idiomatic(mut s: String) -> bool {
        let mut s = Self::clean_string(s.as_str());
        let rev: String = s.chars().rev().collect();
        rev == s
    }
    pub fn is_palindrome(s: String) -> bool {
        let mut s_chars = s
            .chars()
            .filter(|&ch| ch.is_alphanumeric())
            .map(|ch| ch.to_ascii_lowercase());

        while let Some(front) = s_chars.next() {
            if let Some(back) = s_chars.next_back() {
                if front != back {
                    return false;
                }
            } else {
                return true;
            }
        }
        true
    }
    fn clean_string(s: &str) -> String {
        let mut result = String::new();
        for c in s.chars() {
            if c.is_alphanumeric() {
                result.push(c.to_ascii_lowercase());
            }
        }
        result
    }
}

File: python/valid_palindrome_125.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        if len(s) == 0 or len(s) == 1:
            return True
        s = re.sub(r'[^a-zA-Z0-9]', '', s).lower()
        l = 0
        r = len(s) -1
        while l < r:
            if s[l] != s[r]:
                return False
            else:
                l += 1
                r -= 1
        return True
        
